### DISCUSSION TITLE
How to use a Dialog with MVVM

### DISCUSSION THREAD

**Discussion by timunie** [MahApps/MahApps.Metro#4035] [Q&A]
This questions is transferred from #3446

How can we call a Dialog from the `ViewModel`?

**Comment by timunie** [comment]
Hi,

in the demo app you can see a sample how to use the `DialogCoordinator`:
https://github.com/MahApps/MahApps.Metro/blob/bbab412af829c9b565fdfdbfd7aeacd9a30805f3/src/MahApps.Metro.Samples/MahApps.Metro.Demo/MainWindowViewModel.cs#L428-L440

You can add a reference like this: 
https://github.com/MahApps/MahApps.Metro/blob/bbab412af829c9b565fdfdbfd7aeacd9a30805f3/src/MahApps.Metro.Samples/MahApps.Metro.Demo/MainWindowViewModel.cs#L62

https://github.com/MahApps/MahApps.Metro/blob/bbab412af829c9b565fdfdbfd7aeacd9a30805f3/src/MahApps.Metro.Samples/MahApps.Metro.Demo/MainWindowViewModel.cs#L69

And in your `MainWindow`: 
https://github.com/MahApps/MahApps.Metro/blob/bbab412af829c9b565fdfdbfd7aeacd9a30805f3/src/MahApps.Metro.Samples/MahApps.Metro.Demo/MainWindow.xaml.cs#L23-L33

Last but not least in your `MainWindow.xaml`:
https://github.com/MahApps/MahApps.Metro/blob/bbab412af829c9b565fdfdbfd7aeacd9a30805f3/src/MahApps.Metro.Samples/MahApps.Metro.Demo/MainWindow.xaml#L17

--------

More over here is the documentation: https://mahapps.com/docs/dialogs/mvvm-dialog

I hope this helps. 

Happy coding 
Tim

**Reply by msavazzi** [reply]
@timunie sorry for the question, maybe trivial, but I'm confused byt this:

`
         // Set the DataContext for your View
            this.DataContext = new MainWindowViewModel(DialogCoordinator.Instance);
`

but in the app that is create the DataContext is set up automatically in the XAML with:

`
    <Page.DataContext>
        <Binding Path="SettingsViewModel" Source="{StaticResource Locator}" />
    </Page.DataContext>
`

In fact when I add the DialogCoordinator in the method of the view model the XAML binding gives me an error.
What am I doing wrong?

**Reply by timunie** [reply]
@msavazzi You can also access `DialogCoordinator.Instance` directly in your model as far as I remember. Nevertheless you need to register the model in the `XAML`

That means you can remove this Constructor `MainWindowViewModel(DialogCoordinator.Instance);`

Happy coding
Tim

**Reply by msavazzi** [reply]
I did this: In the ShellWindow.xaml I've put:     `Dialog:DialogParticipation.Register="{Binding}"`
The shell window has several pages.
In the MasterDetailPage.xaml the data context is defined as:     `DataContext="{Binding MasterDetailViewModel, Source={StaticResource Locator}}"`
Then in the MasterDetailViewModel the following:

```
        private readonly IDialogCoordinator _dialogCoordinator;
        public MasterDetailViewModel()
        {
            this._dialogCoordinator = DialogCoordinator.Instance;
        }

```
But when I write in the code:
           ` var result = await this.ShowChildWindowAsync<CloseReason>(ChildWindowManager.OverlayFillBehavior.FullWindow);`

I get this error:

> Error	CS1929	'MasterDetailViewModel' does not contain a definition for 'ShowChildWindowAsync' and the best extension method overload 'ChildWindowManager.ShowChildWindowAsync<CloseReason>(Window, ChildWindow, ChildWindowManager.OverlayFillBehavior)' requires a receiver of type 'Window'

	


What am I doing wrong?
I'm very confused

**Reply by timunie** [reply]
`ChildWindow` is not a `Dialog` and therefor cannot be shown via `DialogCoordinator`.
